Firestore/core/src/firebase/firestore/core/sync_engine.mm

```diff
@@ -0,0 +1,497 @@
+
+  for (const auto& entry : query_views_by_query_) {
+    auto query_view = entry.second;
+    const View& view = *query_view->view();
+    ViewDocumentChanges view_doc_changes = view.ComputeDocumentChanges(changes);
+    if (view_doc_changes.needs_refill()) {
+      // The query has a limit and some docs were removed/updated, so we need to
+      // re-run the query against the local store to make sure we didn't lose
+      // any good docs that had been past the limit.
+      DocumentMap docs = [local_store_ executeQuery:query_view->query()];
+      view_doc_changes =
+          view.ComputeDocumentChanges(docs.underlying_map(), view_doc_changes);
+    }
+
+    absl::optional<TargetChange> target_changes;
+    if (maybe_remote_event.has_value()) {
+      const RemoteEvent& remote_event = maybe_remote_event.value();
+      auto it = remote_event.target_changes().find(query_view->target_id());
+      if (it != remote_event.target_changes().end()) {
+        target_changes = it->second;
+      }
+    }
+    ViewChange view_change =
+        query_view->view()->ApplyChanges(view_doc_changes, target_changes);
+
+    UpdateTrackedLimboDocuments(view_change.limbo_changes(),
+                                query_view->target_id());
+
+    if (view_change.snapshot().has_value()) {
+      new_snapshots.push_back(*view_change.snapshot());
+      LocalViewChanges doc_changes = LocalViewChanges::FromViewSnapshot(
+          *view_change.snapshot(), query_view->target_id());
+      document_changes_in_all_views.push_back(std::move(doc_changes));
+    }
+  }
+
+  sync_engine_callback_->OnViewSnapshots(std::move(new_snapshots));
+  [local_store_ notifyLocalViewChanges:document_changes_in_all_views];
+}
+
+void SyncEngine::UpdateTrackedLimboDocuments(
+    const std::vector<LimboDocumentChange>& limbo_changes, TargetId target_id) {
+  for (const LimboDocumentChange& limbo_change : limbo_changes) {
+    switch (limbo_change.type()) {
+      case LimboDocumentChange::Type::Added:
+        limbo_document_refs_.AddReference(limbo_change.key(), target_id);
+        TrackLimboChange(limbo_change);
+        break;
+
+      case LimboDocumentChange::Type::Removed:
+        LOG_DEBUG("Document no longer in limbo: %s",
+                  limbo_change.key().ToString());
+        limbo_document_refs_.RemoveReference(limbo_change.key(), target_id);
+        if (!limbo_document_refs_.ContainsKey(limbo_change.key())) {
+          // We removed the last reference for this key
+          RemoveLimboTarget(limbo_change.key());
+        }
+        break;
+
+      default:
+        HARD_FAIL("Unknown limbo change type: %s", limbo_change.type());
+    }
+  }
+}
+
+void SyncEngine::TrackLimboChange(const LimboDocumentChange& limbo_change) {
+  const DocumentKey& key = limbo_change.key();
+
+  if (limbo_targets_by_key_.find(key) == limbo_targets_by_key_.end()) {
+    LOG_DEBUG("New document in limbo: %s", key.ToString());
```

Comment:
Nit: can you add an empty line after this logging statement? (It's not really related to the rest of the code here)